runtime/compiler/optimizer/TreeLowering.cpp

```diff
@@ -0,0 +1,438 @@
+
+/*******************************************************************************
+ * Copyright (c) 2021, 2021 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "optimizer/TreeLowering.hpp"
+
+#include "compile/Compilation.hpp"
+#include "compile/SymbolReferenceTable.hpp"
+#include "il/Block.hpp"
+#include "il/Block_inlines.hpp"
+#include "infra/ILWalk.hpp"
+
+const char *
+TR::TreeLowering::optDetailString() const throw()
+   {
+   return "O^O TREE LOWERING: ";
+   }
+
+int32_t
+TR::TreeLowering::perform()
+   {
+
```

Comment:
If this opt must be done after GRA, is there any way to add an assert to GRA if this tree lowering has already been performed?

Reply:
The easiest way to add an assert would be to add a flag (probably on the compilation object), set it when this opt runs, and assert on it in GRA. I'm not sure the assert would add much value, but I can add it if others think there would be value in having it.

For clarity, while we believe that there is a performance benefit to running this opt after GRA for the immediate use case (value type lowering), there is no functional need to do so. The requirement I had intended to set is that this opt must be _runnable_ after GRA, meaning that all transformations it performs must handle GlRegDep nodes and such when needed.

Reply:
@0xdaryl any further comments?

Reply:
I'm not sure I like using the Compilation object for that purpose.  But if this is not a strict functional requirement then it likely isn't necessary.